lib/zhuyin.ex

```diff
@@ -0,0 +1,376 @@
+defmodule Zhuyin do
+  @moduledoc """
+  Utilities to deal with zhuyin syllables and groups thereof.
+
+  The main goal of this module is to provide functions to manipulate strings that contain zhuyin
+  words, which are potentially mixed with other content. These strings are represented by the
+  `t:zhuyin_list/0` type. Zhuyin lists can be obtained by parsing a string with the `read/2`,
+  `read!/2` or `sigil_z/2` functions. Afterwards, these lists can be converted into astring
+  representation by using `to_string/1`n.
+
+  A `t:zhuyin_list/0` is a list which contains strings and zhuyin structs (`t:t/0`). These structs
+  are used to encode zhuyin syllables; they can be created directly through the use of the
+  `from_string/1` or `from_string!/1` functions. Like `t:zhuyin_lists/0`, `t:t/0` structs can be
+  converted strings through the use of the `to_string/1`functions.
+
+  Additionally both `t:zhuyin_list/0` and zhuyin structs (`t:t/0`) can be created from or converted
+  to pinyin structs (`Pinyin.t:t/0`) using `from_pinyin/1` or `to_pinyin/1` or
+  """
+
+  alias Zhuyin.Parsers
+
+  # ----- #
+  # Types #
+  # ----- #
+
+  @type t :: %__MODULE__{tone: 0..4, initial: String.t(), final: String.t()}
+
+  @enforce_keys [:final]
+  defstruct tone: 0, initial: "", final: ""
+
+  @typedoc """
+  List of zhuyin syllables mixed with plain strings.
+  """
+  @type zhuyin_list :: [t() | String.t()]
+
+  def zhuyin_tones, do: ["˙", "", "ˊ", "ˇ", "ˋ"]
```

Comment:
It's used down below in the implementation of `defimpl String.Chars, for: Zhuyin do`.
Seems it has to be public for that. Or is there a better way to do it?

Reply:
That's also why it's a funciton and not a variable. Not an Elixir expert, as I said, so it might be stupid.

Reply:
Damn, if I generate `tone_index` at compile-time is can only access variables (not sure if that's the right name) but not the module function. I assume because at that point in time the module isn't available yet.

Reply:
it does work, thanks!

Do you know why `unquote` is required here?